libtenzir/builtins/formats/leef.cpp

```diff
@@ -137,40 +128,56 @@ auto parse_delimiter(std::string_view field) -> std::variant<char, diagnostic> {
   return field[0];
 }
 
-// Converts a raw, unescaped string to a data instance.
-auto to_data(std::string_view str) -> data {
-  auto unescaped = unescape(str);
-  auto result = data{};
-  if (not(parsers::data - parsers::pattern)(unescaped, result)) {
-    result = std::move(unescaped);
-  }
-  return result;
-};
-
 /// Parses the LEEF attributes field as a sequence of key-value pairs.
-auto parse_attributes(char delimiter, std::string_view attributes)
-  -> std::variant<record, diagnostic> {
-  const auto key = +(parsers::printable - '=');
-  const auto value = *(parsers::printable - delimiter);
-  const auto kvp = key >> '=' >> value;
-  const auto kvp_list = kvp % delimiter;
-  auto kvps = std::vector<std::pair<std::string, std::string>>{};
-  if (not kvp_list(attributes, kvps)) {
-    return diagnostic::warning("failed to parse LEEF attributes")
-      .note("attributes: {}", attributes)
-      .done();
-  }
-  auto result = record{};
-  result.reserve(kvps.size());
-  for (auto& [key, value] : kvps) {
-    result.emplace(std::move(key), to_data(value));
+auto parse_attributes(char delimiter, std::string_view attributes,
+                      auto builder) -> std::optional<diagnostic> {
+  while (not attributes.empty()) {
+    auto attr_end = attributes.find(delimiter);
+    auto sep_pos = attributes.find('=');
```

Comment:
What if `sep_pos > attr_end`? You might need a check here  to guard against that, or alternatively, as suggested below: Rewrite to first extract the attribute as a string view and then do all your operations within that.